From: Guillaume Delacour <gui@iroqwa.org>
Subject: Add GNU Hurd support
 * define PATH_MAX
 * Don't use AF_LINK features
 * Read media status with MII and don't include <net/if_media.h>
Last-Update: 2014-08-06

Index: hping3.git/getifname.c
===================================================================
--- hping3.git.orig/getifname.c
+++ hping3.git/getifname.c
@@ -19,7 +19,8 @@
 #include <stdlib.h>
 
 #if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) || \
-    defined(__bsdi__) || defined(__APPLE__) || defined(__FreeBSD_kernel__)
+    defined(__bsdi__) || defined(__APPLE__) || defined(__FreeBSD_kernel__) || \
+    defined(__GNU__)
 #include <ifaddrs.h>
 #include <net/route.h>
 #endif /* defined(__*BSD__) */
@@ -29,7 +30,7 @@
 
 #if !defined(__FreeBSD__) && !defined(__OpenBSD__) && !defined(__NetBSD__) && \
     !defined(__linux__) && !defined(__sun__) && !defined(__bsdi__) && \
-    !defined(__APPLE__) && !defined(__FreeBSD_kernel__)
+    !defined(__APPLE__) && !defined(__FreeBSD_kernel__) && !defined(__GNU__)
 #error Sorry, interface code not implemented.
 #endif
 
@@ -39,7 +40,7 @@
 #include <net/if_dl.h>
 #endif
 
-#if (defined OSTYPE_LINUX) || (defined __sun__)
+#if (defined OSTYPE_LINUX) || (defined __sun__) || defined(__GNU__)
 int get_if_name(void)
 {
 	int fd;
@@ -242,6 +243,8 @@ int get_if_name(void)
 			continue;
 		}
 
+/* GNU Hurd doesn't have AF_LINK support */
+#ifdef AF_LINK
 		if (ifa->ifa_addr->sa_family == AF_LINK) {
 			if (opt_debug)
 				printf("DEBUG: AF_LINK ");
@@ -272,6 +275,7 @@ int get_if_name(void)
 #endif /* __NetBSD__ */
 			continue;
 		}
+#endif /* AF_LINK */
 
 		if (ifa->ifa_addr->sa_family == AF_INET6) {
 			if (opt_debug)
Index: hping3.git/script.c
===================================================================
--- hping3.git.orig/script.c
+++ hping3.git/script.c
@@ -32,6 +32,11 @@
 #include "apdutils.h"
 #include "sbignum.h"
 
+/* for systems that don't have PATH_MAX defined (GNU Hurd) */
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
 #define HPING_IF_MAX	8
 
 /* ----------------------- hping recv handlers code ------------------------- */
@@ -89,7 +94,7 @@ static struct recv_handler *HpingRecvGet
 	if (ra[i].rh_pcapfp == NULL)
 		return NULL;
 	#if (!defined OSTYPE_LINUX) && (!defined __sun__) && \
-		(!defined OSTYPE_GNUKFREEBSD)
+		(!defined OSTYPE_GNUKFREEBSD) && (!defined OSTYPE_GNU)
 	/* Return the packets to userspace as fast as possible */
 	if (ioctl(pcap_fileno(ra[i].rh_pcapfp), BIOCIMMEDIATE, &on) == -1) {
 		/* XXX non-critical error */
@@ -1342,11 +1347,12 @@ static int HpingTcl_AppInit(Tcl_Interp *
 	{
 		char *home = getenv("HOME");
 		if (home) {
-			char rcfile[PATH_MAX];
-			snprintf(rcfile, PATH_MAX, "%s/.hpingrc", home);
-			rcfile[PATH_MAX-1] = '\0';
+			char *rcfile;
+			if (asprintf(&rcfile, "%s/.hpingrc", home) < 0)
+				return TCL_ERROR;
 			Tcl_EvalFile(interp, rcfile);
 			Tcl_ResetResult(interp);
+			free(rcfile);
 		}
 	}
 	return TCL_OK;
Index: hping3.git/libpcap_stuff.c
===================================================================
--- hping3.git.orig/libpcap_stuff.c
+++ hping3.git/libpcap_stuff.c
@@ -35,7 +35,7 @@ int open_pcap()
 		return -1;
 	}
 #if (!defined OSTYPE_LINUX) && (!defined __sun__) && \
-	(!defined OSTYPE_GNUKFREEBSD)
+	(!defined OSTYPE_GNUKFREEBSD) && (!defined OSTYPE_GNU)
 	/* Return the packets to userspace as fast as possible */
 	if (ioctl(pcap_fileno(pcapfp), BIOCIMMEDIATE, &on) == -1)
 		perror("[open_pcap] ioctl(... BIOCIMMEDIATE ...)");
Index: hping3.git/interface.c
===================================================================
--- hping3.git.orig/interface.c
+++ hping3.git/interface.c
@@ -28,16 +28,19 @@
 #include <unistd.h>		/* close */
 
 #if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) || \
-    defined(__bsdi__) || defined(__APPLE__) || defined(__FreeBSD_kernel__)
+    defined(__bsdi__) || defined(__APPLE__) || defined(__FreeBSD_kernel__) || \
+    defined(__GNU__)
 #include <stdlib.h>
 #include <ifaddrs.h>
 #include <net/route.h>
+#if !defined(__GNU__)
 #include <net/if_media.h>
+#endif
 #endif /* defined(__*BSD__) */
 
 #if !defined(__FreeBSD__) && !defined(__OpenBSD__) && !defined(__NetBSD__) && \
     !defined(__linux__) && !defined(__sun__) && !defined(__bsdi__) && \
-    !defined(__APPLE__) && !defined(__FreeBSD_kernel__)
+    !defined(__APPLE__) && !defined(__FreeBSD_kernel__) && !defined(__GNU__)
 #error Sorry, interface code not implemented.
 #endif
 
@@ -61,7 +64,7 @@
  * the interfaces.
  *
  * On error -1 is returned, and errno set. */
-#if (defined OSTYPE_LINUX) || (defined __sun__)
+#if (defined OSTYPE_LINUX) || (defined __sun__) || defined(__GNU__)
 int hping_get_interfaces(struct hpingif *hif, int ilen)
 {
 	int fd, found = 0, i;
